### PR TITLE
shadow ioeventfd: add demo

### PR DESCRIPTION
**Pull request by tmakatos**

### CODE REVIEW

File: samples/shadow_ioeventfd_server.c [hunk 1]
```diff
@@ -0,0 +1,178 @@
...
+ *  CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ *  LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
+ *  OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH
+ *  DAMAGE.
+ *
+ */
+
+/*
+ * shadow_ioeventfd_server.c: an example of how to use a shadow ioeventfd.
+ * There is no Linux kernel driver, use samples/shadow_ioevetnfd_client.c in
```
Comment (jlevon) [inline samples/shadow_ioeventfd_server.c:32]: typo

Comment (jlevon) [reply samples/shadow_ioeventfd_server.c:32]: it's not called client.c either any more

File: samples/shadow_ioeventfd_server.c [hunk 2]
```diff
@@ -0,0 +1,178 @@
...
+#include <assert.h>
+#include <sys/poll.h>
+#include <sys/eventfd.h>
+#include <sys/mman.h>
+#include <sys/syscall.h>
+
+#include "libvfio-user.h"
+
+static void
+_log(vfu_ctx_t *vfu_ctx __attribute__((unused)),
```
Comment (jlevon) [inline samples/shadow_ioeventfd_server.c:49]: here, everywhere, use UNUSED

File: samples/shadow_ioeventfd_server.c [hunk 3]
```diff
@@ -0,0 +1,178 @@
...
+
+int
+main(int argc, char *argv[])
+{
+    int ret;
+    vfu_ctx_t *vfu_ctx;
+    struct pollfd fds[2]; /* one for vfu_ctx, one for shadow_ioeventfd */
+    int fd, bar0_fd;
+
+    if (argc != 2 ) {
```
Comment (jlevon) [inline samples/shadow_ioeventfd_server.c:74]: spurious space

File: samples/shadow_ioeventfd_speed_test.c [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+	int ret = ioctl(group, VFIO_GROUP_GET_STATUS, &status);
+	assert(ret != -1);
+	assert(status.flags & VFIO_GROUP_FLAGS_VIABLE);
+	ret = ioctl(group, VFIO_GROUP_SET_CONTAINER, &container);
+	assert(ret != -1);
+	ret = ioctl(container, VFIO_SET_IOMMU, VFIO_TYPE1_IOMMU);
+	assert(ret == 0);
+	int device = ioctl(group, VFIO_GROUP_GET_DEVICE_FD, argv[2]);
+	assert(device >= 0);
+	assert(device != -1);
```
Comment (jlevon) [inline samples/shadow_ioeventfd_speed_test.c:69]: no need for this

**Review by jlevon — APPROVED** [review]
thanks

**Comment by tmakatos** [comment]
I already merged this into [multiple-shadow-ioeventfd](https://github.com/nutanix/libvfio-user/tree/multiple-shadow-ioeventfd) but I should have waited until the destination branch had been merged into master and then merge.

**Pull request merged**